src/storage-client/src/controller.rs

```diff
@@ -1738,170 +1757,143 @@ mod persist_read_handles {
-                                                let fut = async move {
-                                                    // Use maybe_downgrade_since here so that we opt
-                                                    // into rate-limiting. It's okay for the since to
-                                                    // lag behind a bit and this _greatly_ reduces the
-                                                    // persist traffic.
-                                                    read.maybe_downgrade_since(&since).instrument(span).await;
-                                                };
+                                let snapshot = read_handle
+                                    .snapshot_and_fetch(as_of)
+                                    .instrument(span.clone())
+                                    .await;
+                                let result = match snapshot {
+                                    Ok(contents) => Ok(contents
+                                        .into_iter()
+                                        .map(|((data, _), _, diff)| {
+                                            (data.expect("invalid protobuf data"), diff)
+                                        })
+                                        .collect()),
+                                    Err(_) => Err(StorageError::ReadBeforeSince(id)),
+                                };
 
-                                                futs.push(fut);
-                                            }
-                                        }
-                                    }
+                                read_handle.expire().await;
 
-                                    assert!(downgrades.is_empty());
-                                    futs.collect::<Vec<_>>().await;
+                                if oneshot.send(result).is_err() {
+                                    tracing::warn!("snapshot requested but requester is gone");
                                 }
+                            }
+                        }
+                    }
 
-                                // This should be pretty rare, so don't bother doing them all in a
-                                // FuturesUnordered.
-                                for id in drops {
-                                    let read = read_handles.remove(&id).unwrap().unwrap();
-                                    read.expire().await;
-                                }
+                    let mut futs = FuturesUnordered::new();
+
+                    for (id, (span, since)) in downgrades {
+                        let Some((metadata, mut since_handle)) = collection_state.remove(&id) else {
+                            panic!("downgrade command for absent collection {id}");
+                        };
+
+                        futs.push(async move {
+                            since_handle
+                                .maybe_compare_and_downgrade_since(&0, (&0, &since))
```

Comment:
IIUC the token here is meant to be the envd epoch, but you'll want to talk to @jkosh44 about where that lives and how we should plumb it down (probably can be a constructor arg to `persist_read_handles::PersistWorker`)

Reply:
Currently environmentd creates the `mz_controller::Controller`, with the epoch here:
https://github.com/MaterializeInc/materialize/blob/aaf3558f4e983cf5b432bcd2af35982cdc0c2c89/src/environmentd/src/lib.rs#L278-L279

The `mz_controller::Controller` creates the `mz_storage_client::controller::Controller` here: https://github.com/MaterializeInc/materialize/blob/aaf3558f4e983cf5b432bcd2af35982cdc0c2c89/src/controller/src/lib.rs#L213-L224

The `mz_storage_client::controller::Controller` creates the `StorageControllerState` here:
https://github.com/MaterializeInc/materialize/blob/aaf3558f4e983cf5b432bcd2af35982cdc0c2c89/src/storage-client/src/controller.rs#L1381-L1408

and finally the `StorageControllerState` creates the `persist_read_handles::PersistWorker` here:
https://github.com/MaterializeInc/materialize/blob/aaf3558f4e983cf5b432bcd2af35982cdc0c2c89/src/storage-client/src/controller.rs#L738-L771

So you should be able to easily plumb the epoch down through all those constructors if I followed everything correctly.

Reply:
If you do need to access the epoch in other places, it's actually a bit hidden. It's made available through the stash API here: 
https://github.com/MaterializeInc/materialize/blob/aaf3558f4e983cf5b432bcd2af35982cdc0c2c89/src/stash/src/lib.rs#L396-L398

You'll likely need to add a method to the `Connection` object that exposes this:
https://github.com/MaterializeInc/materialize/blob/aaf3558f4e983cf5b432bcd2af35982cdc0c2c89/src/adapter/src/catalog/storage.rs#L529
Something like 
```Rust
    pub fn epoch(&self) -> Result<Option<i64>, Error> {
        Ok(self.stash.epoch().await?)
    }
```

Then you'll be able use that method in the Coordinator to pass it to the storage controller/client like so:
```Rust
pub async fn some_coord_method(&self) {
  let epoch = self.catalog.storage().await.epoch()?;
  // do something with epoch.
}
```

Reply:
This was super helpful, thanks! I will wire it up tomorrow